File: 345. Reverse Vowels of a String.java
/**
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 */
public class ReverseVowelsOfAString {
    // O(n) time complexity (n=s.length)
    // O(n) space complexity (n=s.length)
    String reverseVowels(String s) {
        int l = 0;
        int r = s.length() - 1;

        char[] arr = s.toCharArray();

        while (l < r) {
            char cl = s.charAt(l);
            char cr = s.charAt(r);
            if (isVowel(cl) && isVowel(cr)) {
                char temp = arr[l];
                arr[l] = arr[r];
                arr[r] = temp;
                l++;
                r--;
            } else if (isVowel(cl)) {
                r--;
            } else if (isVowel(cr)) {
                l++;
            } else {
                l++;
                r--;
            }
        }
        return new String(arr);
    }

    private boolean isVowel(char c) {
        c = Character.toLowerCase(c);
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
}

class M {
    public static void main(String[] args) {
        ReverseVowelsOfAString r = new ReverseVowelsOfAString();
        String input = "leetcode";
        System.out.println(r.reverseVowels(input));
    }
}

File: 345. Reverse Vowels of a String.kt
/**
 * https://leetcode.com/problems/remove-vowels-from-a-string/
 */

// O(n) time complexity (n=s.length)
// O(n) space complexity (n=s.length)
fun reverseVowels(s: String): String {
    var l = 0
    var r = s.length - 1
    val arr = s.toCharArray()
    while (l < r) {
        if (isVowel(arr[l]) && isVowel(arr[r])) {
            val temp = arr[l]
            arr[l] = arr[r]
            arr[r] = temp
            l++
            r--
        } else if (isVowel(arr[l])) {
            r--
        } else if (isVowel(arr[r])) {
            l++
        } else {
            l++
            r--
        }
    }
    return String(arr)
}

fun isVowel(input: Char): Boolean {
    return input == 'a' || input == 'e' || input == 'i' || input == 'o' || input == 'u'
            || input == 'A' || input == 'E' || input == 'I' || input == 'O' || input == 'U'
}

fun main() {
    val input = "leetcode"
    println(reverseVowels(input)) // output: leotcede
}